src/execution/execute.ts

```diff
@@ -1491,87 +1491,134 @@ function deferredFragmentRecordFromDeferUsage(
   deferUsage: DeferUsage,
   deferMap: ReadonlyMap<DeferUsage, DeferredFragmentRecord>,
 ): DeferredFragmentRecord {
-  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-  return deferMap.get(deferUsage)!;
+  const deferredFragmentRecord = deferMap.get(deferUsage);
+  invariant(
+    deferredFragmentRecord != null,
+    'deferredFragmentRecord must be in deferMap',
+  );
+  return deferredFragmentRecord;
 }
 
-function addNewDeferredGroupedFieldSets(
-  incrementalPublisher: IncrementalPublisher,
-  newGroupedFieldSetDetails: Map<DeferUsageSet, GroupedFieldSetDetails>,
-  deferMap: ReadonlyMap<DeferUsage, DeferredFragmentRecord>,
-  path?: Path | undefined,
-): ReadonlyArray<DeferredGroupedFieldSetRecord> {
-  const newDeferredGroupedFieldSetRecords: Array<DeferredGroupedFieldSetRecord> =
-    [];
-
-  for (const [
-    newGroupedFieldSetDeferUsages,
-    { groupedFieldSet, shouldInitiateDefer },
-  ] of newGroupedFieldSetDetails) {
-    const deferredFragmentRecords = getDeferredFragmentRecords(
-      newGroupedFieldSetDeferUsages,
-      deferMap,
-    );
+function getGroupedFieldSetsByTarget(
+  exeContext: ExecutionContext,
+  groupedFieldSet: GroupedFieldSet,
+  path: Path | undefined,
+  incrementalDataRecord: IncrementalDataRecord,
+  deferMap?: ReadonlyMap<DeferUsage, DeferredFragmentRecord>,
+): {
+  currentGroup: GroupedFieldSet;
+  deferredGroupedFieldSetRecords: Array<DeferredGroupedFieldSetRecord>;
+  newDeferMap: ReadonlyMap<DeferUsage, DeferredFragmentRecord>;
+} {
+  const currentGroup = new Map<string, ReadonlyArray<FieldDetail>>();
+  const newDeferUsages = new Set<DeferUsage>();
+  const deferredGroups: Array<{
+    deferUsageSet: Set<DeferUsage>;
+    shouldInitiateDefer: boolean;
+    groupedFieldSet: GroupedFieldSet;
+  }> = [];
+
+  for (const [key, fieldDetails] of groupedFieldSet) {
+    const deferUsageSet = new Set<DeferUsage>();
+    let everyNodeIsDeferred = true;
+    for (const fieldDetail of fieldDetails) {
+      if (fieldDetail.deferUsage != null) {
+        deferUsageSet.add(fieldDetail.deferUsage);
+        newDeferUsages.add(fieldDetail.deferUsage);
+      } else {
+        everyNodeIsDeferred = false;
+      }
+    }
+    const parentDeferUsageSet: Set<DeferUsage> =
+      'deferredFragmentRecords' in incrementalDataRecord
+        ? new Set(
+            incrementalDataRecord.deferredFragmentRecords.map(
+              (d) => d.deferUsage,
+            ),
+          )
+        : new Set();
+    if (
+      !everyNodeIsDeferred ||
+      (parentDeferUsageSet.size !== 0 &&
+        setIsSubsetOf(parentDeferUsageSet, deferUsageSet))
+    ) {
+      currentGroup.set(key, fieldDetails);
+    } else {
+      let deferredGroup = deferredGroups.find((group) =>
+        isSameSet(group.deferUsageSet, deferUsageSet),
+      );
```

Comment:
because masking here is done differently, this find is not necessarily stable, consider the operation:

```graphql
query HeroNameQuery {
        ... @defer(label: "DeferID") {
          hero {
            id
          }
        }
        ... @defer(label: "DeferName") {
          hero {
            name
            lastName
            ... @defer {
              lastName
            }
          }
        }
      }
```

`hero` is used in two parallel defers, with the `id` subfield present in `DeferID` and both `name` and `lastName` present in `DeferName`. The subfields of `hero` should therefore yield two new grouped field sets, but according to this algorithm, it would yield three, as the additional superfluous defer would still be counted and so the `deferUsageSet` for `lastName` would actually be different for `name`.

Reply:
to fix this, you may need to bring back `ancestors` on `DeferUsage`